build_tools/cmake/build_all.sh

```diff
@@ -54,6 +54,10 @@ declare -a CMAKE_ARGS=(
   "-DIREE_HAL_DRIVER_CUDA=ON"
   "-DIREE_TARGET_BACKEND_CUDA=ON"
 
+  # Enable WebGPU compiler builds and tests. All deps get fetched as needed,
+  # but some of the deps are too large to enable by default for all developers.
+  "-DIREE_TARGET_BACKEND_WEBGPU=ON"
```

Comment:
@GMNGeoffrey / @pzread FYI - if you want to review this part of the PR (covered by CODEOWNERS :P)

Reply:
I also mention this a bit in my RFC: [Promoting IREE web platform / Emscripten builds to "stable"](https://groups.google.com/g/iree-discuss/c/2K5VJ9P8K8I/m/elidhfFMCAAJ)

Enabling this option fetches Tint on-demand for the WebGPU compiler target to use to compile SPIR-V to WGSL.

Reply:
Just synced this PR and I'm looking to land it. @GMNGeoffrey does this satisfy your point on the RFC?

> Regarding on-demand dependencies. Can we make these as not-magic as possible?

https://github.com/iree-org/iree/blob/fe40a3db1ee9617e8daa9ae2d3e729af603ad8a7/CMakeLists.txt#L702-L709

https://github.com/iree-org/iree/blob/main/build_tools/third_party/tint/CMakeLists.txt

Reply:
Oh dear... maybe these are filling up the runner's disk space? https://github.com/iree-org/iree/actions/runs/3137979962

```
build_all
Unhandled exception. System.IO.IOException: No space left on device : '/runner-root/actions-runner/_diag/Worker_20220927-182608-utc.log'
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
System.IO.IOException: No space left on device : '/runner-root/actions-runner/_diag/Worker_20220927-182608-utc.log'
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
   at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.Diagnostics.TextWriterTraceListener.Flush()
   at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id)
   at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message)
   at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message)
   at GitHub.Runner.Worker.Worker.RunAsync(String pipeIn, String pipeOut)
   at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args)
System.IO.IOException: No space left on device : '/runner-root/actions-runner/_diag/Worker_20220927-182608-utc.log'
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
   at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.Diagnostics.TextWriterTraceListener.Flush()
   at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id)
   at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.Diagnostics.TextWriterTraceListener.Flush()
   at System.Diagnostics.TraceSource.Flush()
   at GitHub.Runner.Common.TraceManager.Dispose(Boolean disposing)
   at GitHub.Runner.Common.TraceManager.Dispose()
   at GitHub.Runner.Common.HostContext.Dispose(Boolean disposing)
   at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message)
   at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message)
   at GitHub.Runner.Common.Tracing.Error(Exception exception)
   at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args)
   at GitHub.Runner.Common.HostContext.Dispose()
   at GitHub.Runner.Worker.Program.Main(String[] args)
```

Reply:
Maybe we should increase the CPU runner size? https://github.com/iree-org/iree/pull/10518

https://github.com/iree-org/iree/blob/5bf7ae901980a3180b9067fd52a815a48122eb44/build_tools/github_actions/runner/gcp/create_templates.sh#L26-L28

Can we view a breakdown of how disk space is being used?

Reply:
Sorry, I missed the original mention here. I may need to adjust my email filters to differentiate "assigned as reviewer because someone touched a build config" and "someone explicitly mentioned me"...

So the runner is actually doing its work in a 50G ramdisk (you reviewed that PR :stuck_out_tongue:): https://github.com/iree-org/iree/blob/5c1ce206c2/build_tools/github_actions/runner/config/setup.sh#L19. Until now that was plenty of space for building IREE. How big is this dependency exactly?? I don't love fetching on-demand on CI. For CUDA we build the necessary parts of the dep into the docker image. Can we do that instead?

Separately: turning this on here puts building the Web GPU backend in everyone's presubmits. Is it sufficiently stable for that?

Reply:
> How big is this dependency exactly??

The FetchContent `_deps/tint-*` and `_deps/spirv-tools-*` folders are 1.7GB

> For CUDA we build the necessary parts of the dep into the docker image. Can we do that instead?

Not... really. This is a source dependency on a library, and these projects are not distributed as an SDK like with CUDA.

> Separately: turning this on here puts building the Web GPU backend in everyone's presubmits. Is it sufficiently stable for that?

The WebGPU compiler target should be stable yes. Outside of the new dep on Tint, the compiler target is very small - it's mostly SPIRV + a few extra passes + one new linker tool. The runtime HAL is still unstable and under development.

Reply:
50G feels a bit small for building IREE (my source + build dirs combined are ~40GB on Windows, which doesn't give us much headroom). Would increasing that to 100GB be reasonable?

Reply:
The GPU machines only have 85GB of memory. I actually was only looking at the CPU machines when I decided to just use a ramdisk, which have 360GB. I didn't realize we were running so close to the limit on the GPU machines. We should probably switch to a local SSD

Reply:
Are we actually building on GPU machines, or just running already built tests/benchmarks/etc.? (The new deps here should only affect the compiler build on CPU/host machines)

Reply:
We're not building, but the setup step for the ramdisk is the same for both. We could add a conditional. But I actually think using 50/85 GB on the ramdisk is probably a bad idea anyway. We could see about lowering the also

Reply:
Hmmmm... I'd still like to land this soon, so I can rebase the `benvanik-webgpu-skeleton` branch on the compiler changes here. I think I'll split this into two PRs: one updating the code, and one turning on the CI. The CI change will need to either change how the deps are installed or wait for infrastructure changes (more disk space, or more efficient use of the current disk space).

Reply:
SGTM

Reply:
Moved the CI-affecting changes to https://github.com/iree-org/iree/pull/10627